src/app/clusters/mode-base-server/mode-base-server.h

```diff
@@ -0,0 +1,198 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "mode-base-cluster-objects.h"
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/util/af.h>
+#include <map>
+
+using chip::Protocols::InteractionModel::Status;
+using ModeOptionStructType = chip::app::Clusters::detail::Structs::ModeOptionStruct::Type;
+using ModeTagStructType    = chip::app::Clusters::detail::Structs::ModeTagStruct::Type;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ModeBase {
+
+class Instance : public CommandHandlerInterface, public AttributeAccessInterface
+{
+public:
+    // This map holds pointers to all initialised ModeBase instances. It provides a way to access all ModeBase clusters.
+    static std::map<uint32_t, Instance *> ModeBaseAliasesInstanceMap;
```

Comment:
- Please have this be a per-instance --> not static. The fact there is a single instance will smooth that over. Please add a getter for it by const ref instead, and a method to register/deregister a new instance (rather than getter by ref + mutating it at caller site).
- std::map may not be supported on all platforms, and is costly to insert unless reservations are made. Let's keep it like this for now, with a view to refactor away from <map>.
- Consider std::vector of a struct having <uint32_t, Instance *> and linear search rather than std::map --> less flash, less RAM, and very efficient for small number of registered instances (the common case)

Reply:
- Not sure what is meant by "per-instance". This is a store that contains all active ModeBase alias cluster instances, so it wouldn't really work per-instance. This singleton way of doing things is to deal with the limitations of the OnOff server. I have logged an issue to fix this [here](https://github.com/project-chip/connectedhomeip/issues/27508). For the time being, I have moved this into the ModeBase namespace and added a getter. This is still not const as otherwise I would have to return a single element at a time which will add extra complication to this temporary code.
- I have been told previously not to use `std::vector` in embedded code. Can anyone recommend an embedded friendly way to build a map? Some kind of a key-value store is going to be required to allow for de-registration.

Reply:
This has now been changed to use `std::set`.

Reply:
map/set/vector are basically all the same: they are STL and  unfriendly to embedded devices, both in terms of flash space (since it's large templates) and runtime heap/exceptions behavior. For now, you can keep it like this an optimize later